cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,929 @@
+                         edge_t{num_valid_edges},
+                         raft::device_span<edge_t>(num_triangles.data(), num_triangles.size()),
+                         get_dataframe_buffer_begin(vertex_pair_buffer_p_r_edge_p_q),
+                         transposed_edge_first,
+                         transposed_edge_first + num_valid_edges,
+                         transposed_edge_first + edgelist_srcs.size()});
+
+      auto vertex_pair_buffer_q_r_edge_p_q =
+        allocate_dataframe_buffer<thrust::tuple<vertex_t, vertex_t>>(accumulate_pair_size,
+                                                                     handle.get_stream());
+
+      thrust::tabulate(
+        handle.get_thrust_policy(),
+        get_dataframe_buffer_begin(vertex_pair_buffer_q_r_edge_p_q),
+        get_dataframe_buffer_end(vertex_pair_buffer_q_r_edge_p_q),
+        generate_p_r_q_r<vertex_t, edge_t, false>{
+          raft::device_span<size_t const>(intersection_offsets.data(), intersection_offsets.size()),
+          raft::device_span<vertex_t const>(intersection_indices.data(),
+                                            intersection_indices.size()),
+          raft::device_span<vertex_t>(edgelist_srcs.data() + num_valid_edges, num_invalid_edges),
+          raft::device_span<vertex_t>(edgelist_dsts.data() + num_valid_edges, num_invalid_edges)});
+
+      thrust::for_each(handle.get_thrust_policy(),
+                       thrust::make_counting_iterator<edge_t>(0),
+                       thrust::make_counting_iterator<edge_t>(num_edge_exists),
+                       unroll_edge<vertex_t, edge_t, decltype(transposed_edge_first)>{
+                         edge_t{num_valid_edges},
+                         raft::device_span<edge_t>(num_triangles.data(), num_triangles.size()),
+                         get_dataframe_buffer_begin(vertex_pair_buffer_q_r_edge_p_q),
+                         transposed_edge_first,
+                         transposed_edge_first + num_valid_edges,
+                         transposed_edge_first + edgelist_srcs.size()});
+
+      // case 1: unroll (q, r)
+      // For each (q, r) edges to unroll, find the incoming edges to 'r' let's say from 'p' and
+      // create the pair (p, q)
+      cugraph::find_unroll_p_r_and_q_r_edges<vertex_t, edge_t, false, true>(
+        handle,
+        cur_graph_view,
+        num_invalid_edges,
+        num_valid_edges,
+        std::move(edgelist_srcs),
+        std::move(edgelist_dsts),
+        std::move(num_triangles));
+
+      // case 2: unroll (p, r)
+      cugraph::find_unroll_p_r_and_q_r_edges<vertex_t, edge_t, false, false>(
+        handle,
+        cur_graph_view,
+        num_invalid_edges,
+        num_valid_edges,
+        std::move(edgelist_srcs),
+        std::move(edgelist_dsts),
+        std::move(num_triangles));
+
+      cur_graph_view.clear_edge_mask();
+      edges_with_no_triangle.clear();
+      edge_mask.clear(handle);  // masking not working in a loop.
```

Comment:
Need to resize edgelist_srcs, edgelist_dsts, triangle_counts and mask out the invalid edges here, right?

Reply:
I am already doing this at the beginning of the loop. I am not sure why I would I do it here?